Image editor

pngwn opened the pull request:
## user facing

This is the new `ImageEditor`, which is a replacement for 3.x's `Image` with `tool=...`. I have linked the issues that it fixes below (both bugs and enhancements), so i wont go over every detail but the high level is as follows:

This component is a simple + streamlined image editor that handles basic image manipulations. It should be at least equivalent to the old Image with various `tools` set but less bad and more good. Additional features include:

- **Layers**
  - The new component had a concept of layers. Certain operations are constrained to a single layer and do not impact other layers. Details on each tool.
  - Each layers is returned individually to the user function. 

- **Background**
  - has a concept of backgrounds. 
  - A background is a special type of layer than cannot be drawn on or erased.
  - default is a plain white background which can be drawn on.
  - can also accept an image as the 'backgrounds'
    - `upload` will upload a file from the users computer 
    - `webcam` will open a webcam capture window which will be set as the bg when it is captured.
    - `paste` will paste from clipboard
- **Draw**
  - users can draw on top of any background.
  - drawing is constrained to a single layer and does impact others.
  - The size and color can be selected by the user from a set of swatches and optionally an arbitrary 'picker'
  - The picker is only present if the app author allows it.
  - The size picker is a range input (slider)
  - The color picker is a custom color picker that users will be somewhat familiar with. The color picker also allows users to build their own swatches.
  - I think the color picker is very nice.
- **Erase**
  - Erase is basically draw but erase.
  - It only affects drawings on the currently selected layer
  - It has only size  which operates the same as draw's size
- **Crop**
  - Crops the entire canvas 
  - Crop is non destructive, you can re-enter crop mode and uncrop however you like.
  - i think crop is very nice too.
- **Undo**
  - as in 3.x we have undo, except this time is actually works 
- **Redo**
  - Because of the architecture (or the fact that it even has one), we can also easily 'redo' any operation.
- **Performance**
  - The new image editor is much more performant in my experience. It seems to work well with pretty large images and lots of operations. There are definitely opportunities to optimise though, history is not cheap.

## technical

The new Image Editor is built on top of [pixi.js](https://pixijs.com/), a javascript library for creating 2d webgl experiences. It is an exceptional library if a little confusing but it sits at an almost perfect level of abstraction. High level enough so you don't need to think too much about the inner workings, low level enough that you can dive in when you need to. It is not a sketch library in any way but has all of the correct primitives. It is also really fast.

The image editor has a bunch of code but its a lot less than I thought it would need to be. The image editor has been designed to be flexible and maintainable, it should be relatively easy to add new functionality without needing to touch the whole thing. For example, modification to the crop tool, should only require changes to the cropping part of the code.

The editor uses the [command pattern](https://en.wikipedia.org/wiki/Command_pattern) which is a little complex but incredibly powerful. Every editing action is a 'command'. A command knows how to perform itself and how to undo itself. We have a command manager that takes in every command, and adds it to its history. When we want a new feature, we create a new command that is dedicated to just that thing. `crop` is a command, `draw` is a command, 'add bg image` is a command.

When we want to undo something we ask the command manager to undo it, it will then call the current history node's `undo` method and move a point back. And repeat.

When we want to redo something it will call the 'next' commands `execute` method, and move the history point forwards. The history is implement as a doubly linked list so it is simple to traverse and very difficult to skip nodes.

The 'command' interfaces look like this:

```ts
interface Command {
  start?: (...args: any[]) => any;
  continue?: (...args: any[]) => any;
  stop?: (...args: any[]) => any;
  execute: () => void;
  undo: () => void
}
```

Each command can decide how to implement all of these methods, it doesn't matter as long as they 'work'. Only `execute` and `undo` are required.

The reason `start` `continue` and `stop` exist is best explained with an example:

Lets take drawing. A single drawing operation is one line. So `execute` should draw an entire line onto the canvas (for example when we are 'redoing' a line). However, we can only call execute when the line has finished drawing because that will get added to the history as a single entity. Since this is an image editor we need real time updates. With this in mind, i added `start`, `continue`, and `stop` as optional methods that can be used to provide realtime updates to the canvas, while still giving us the option to undo/redo later.

In the case of drawing:
- we create a new command and trigger `start` when a user presses their mouse down. We set up some initial textures and snapshot the initial state. 
- When the user moves their mouse we trigger `continue`. here we are interpolating between x/y points (to ensure a smooth line), drawing to the texture we set up before and rendering it.
- When the user releases their mouse button, we call `stop`. In stop we perform some final clean up and reset the textures. We reuse the textures for performance reasons, so this clean up is important.

After realtime drawing `execute` is basically a no-op (it always get called when we pass it to the command manager). However if no realtime drawing has been completed then it will draw the full line.

The reason it is important that the command manager is able to handle these realtime updates is because all drawing operations then just live in this one ovject which makes maintenance much easier. Since everything is passed into these commands as arguments, unit testing is also straight forward.

The command pattern is actually really nice for this use case and allows us to easily add complex features without drowning in a sea of complexity. It also unlock other possible features in the future (really fancy ones like macros/ batch processing).
 

**jsdoc**

I have made extensive use of jsdoc annotations throughout the code in the hope that it offers a  little help for us all when navigating this component in the future.

This is a practice that I think we should experiment with across the rest of the code base (frontend). I personally find it very helpful, both when writing and reading code because it makes me think about my interfaces + what they are actually doing.


Closes #5645 
Closes #5206 
Closes #5152 
Closes #5055 
Closes #4931 
Closes #4907 
Closes #4842 
Closes #4677 
Closes #4653 
Closes #4492 
Closes #4413 
Closes #4290 
Closes #4252 
Closes #4159 
Closes #4120
Closes #4011
Closes #3810 
Closes #3623 
Closes #3535
Closes #3472 
Closes #3305 
Closes #3280 
Closes #3138 
Closes #3110 
Closes #2649 
Closes #2425 
Closes #2314 
Closes #1591 
Closes #466 

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

abidlabs commented on gradio/components/image_editor.py:
```diff
@@ -0,0 +1,297 @@
+_Image.init()  # fixes https://github.com/gradio-app/gradio/issues/2843
+
+
+class PreprocessData(TypedDict):
```
Nice, its cool to have a typeddict to store the value as well. I find `PreprocessData` a bit confusing name, since its also the format that a user is expected to return from their function. How about `EditorValue` as it should be the same type as `value`

abidlabs commented on gradio/components/image_editor.py:
```diff
@@ -0,0 +1,297 @@
+    composite: Optional[FileData] = None
+
+
+class Eraser:
```
Nice

abidlabs commented:
@pngwn I'm happy to review this, but it looks like the frontend code may have broken as a result of the v4 changes, and it can't build at the moment. 

Reviewed the backend, looks very good. I'm excited about the Brush and Eraser classes, and will try to recreate some of the old demos (like the sketchpad demo) with this.

pngwn commented:
hmm

freddyaboulton commented on gradio/components/image_editor.py:
```diff
@@ -0,0 +1,297 @@
+
+class EditorData(GradioModel):
+    background: Optional[Union[FileData, str]] = None
+    layers: list[FileData] = []
```
Use List[FileData]

pngwn commented on .changeset/few-tips-appear.md:
```diff
@@ -0,0 +1,11 @@
+"gradio": minor
+---
+
+feat:Image editor
```
Update this with a highlight.

pngwn commented on demo/image_editor/run.py:
```diff
@@ -0,0 +1,24 @@
+    btn = gr.Button()
+    # btn.click(image, outputs=[im2, im3, im4, im5], inputs=im)
+    btn.click(lambda x: x, outputs=im_out, inputs=im)
+
```
Expand demo. Uodate image mask demo. Add to deploy.

abidlabs replied:
It would be great to have a more complex demo as well, that shows e.g. a custom `Brush`, `crop_size`, etc.

pngwn commented on gradio/components/image_editor.py:
```diff
@@ -0,0 +1,302 @@
+        return str(utils.abspath(input_data))
+
+    def example_inputs(self) -> Any:
+        return "https://raw.githubusercontent.com/gradio-app/gradio/main/test/test_files/bus.png"
```
Need to bring @freddyaboulton's recent fixes across.

pngwn commented on js/app/src/Blocks.svelte:
```diff
@@ -19,6 +19,7 @@
 
 	import logo from "./images/logo.svg";
 	import api_logo from "./api_docs/img/api-logo.svg";
+	import type { P } from "@storybook/svelte-vite/dist/index-55234481";
```
Remove.

pngwn commented on js/app/src/Blocks.svelte:
```diff
@@ -374,6 +375,13 @@
 	let showed_duplicate_message = false;
 	let showed_mobile_warning = false;
 
+	function get_data(comp: ComponentMeta): any | Promise<any> {
```
Document somewhere. Need to add the (change) event detection too.

pngwn commented on js/image/shared/Webcam.svelte:
```diff
@@ -15,29 +15,25 @@
 		try {
 			stream = await navigator.mediaDevices.getUserMedia({
-				video: device_id ? { deviceId: { exact: device_id } } : true,
+				video: device_id ? { deviceId: { exact: device_id }, ...size } : size,
```
We can now capture much higher resolution webcam images.

pngwn commented on js/imageeditor/CHANGELOG.md:
```diff
@@ -0,0 +1,131 @@
+# @gradio/image
```
Clear this file.

pngwn commented on js/imageeditor/Example.svelte:
```diff
@@ -0,0 +1,39 @@
+<script lang="ts">
+	import Image from "./shared/Image.svelte";
+
+	export let value: string;
```
This is wrong, it's an object.

pngwn commented on js/imageeditor/Example.svelte:
```diff
@@ -0,0 +1,39 @@
+	class:gallery={type === "gallery"}
+	class:selected
+>
+	<Image src={samples_dir + value} alt="" />
```
We should show the composite here.

pngwn commented on js/imageeditor/Index.svelte:
```diff
@@ -0,0 +1,175 @@
+<svelte:options accessors={true} />
+
+<script context="module" lang="ts">
+	// export { default as Webcam } from "./shared/Webcam.svelte";
```
Remove.

pngwn commented on js/imageeditor/Index.svelte:
```diff
@@ -0,0 +1,175 @@
+			{eraser}
+			i18n={gradio.i18n}
+		>
+			<!-- {#if sources.includes("upload")}
+				<UploadText i18n={gradio.i18n} type="image" mode="short" />
+			{:else}
+				<Empty unpadded_box={true} size="large"><Image /></Empty>
+			{/if} -->
```
Remove.

pngwn commented on js/imageeditor/README.md:
```diff
@@ -0,0 +1,43 @@
+# `@gradio/imageeditor`
```
Update.

pngwn commented on js/imageeditor/package.json:
```diff
@@ -0,0 +1,30 @@
+		"@gradio/wasm": "workspace:^",
+		"@types/tinycolor2": "^1.4.6",
+		"pixi.js": "^7.3.2",
+		"resize-observer-polyfill": "^1.5.1",
```
Remove.

pngwn commented on js/imageeditor/shared/Image.svelte:
Remove.

pngwn commented on js/imageeditor/shared/ImagePreview.svelte:
I think we can remove the static variant and just import from `@gradio/image` instead.

pngwn commented on js/imageeditor/shared/layers/Layers.svelte:
```diff
@@ -0,0 +1,193 @@
+
+	$: $pixi && once_layer();
+
+	async function new_layer(): Promise<void> {
+		if (!$pixi) return;
+
+		await tick();
+
+		setTimeout(() => {
+			const [active_layer, all_layers] = LayerManager.add_layer(
+				$pixi.layer_container,
+				$pixi.renderer,
+				...$dimensions
+			);
+
+			$current_layer = active_layer;
+			layers = all_layers;
+		}, 20);
+	}
```
This is bad. We need to figure something else out.

pngwn commented on js/imageeditor/shared/layers/Layers.svelte:
```diff
@@ -0,0 +1,193 @@
+		layers = last_layer[1];
+	}
+
+	onMount(async () => {
+		await tick();
+		if (!$pixi) return;
+		function reset(): void {
+			console.log("resetting");
+			LayerManager.reset();
+			$pixi?.resize(...$dimensions);
+
+			new_layer();
+		}
+		$pixi = { ...$pixi!, reset, get_layers: LayerManager.get_layers };
+	});
```
I think this is the problem. We do this in a few places and it suck's because we can't guarantee the order that the various reset functions are called. 

Instead we should dispatch events from the core component when we want to reset something and manually call reset at a higher level. Alternatively we can track functions more granularly.

pngwn commented on js/imageeditor/shared/layers/Layers.svelte:
```diff
@@ -0,0 +1,193 @@
+			{#each layers as layer, i (i)}
+				<li>
+					<button
+						class:selected_layer={$current_layer === layer}
```
Add some controls to manipulate layers.

pngwn commented on js/imageeditor/shared/layers/Layers.svelte:
```diff
@@ -0,0 +1,193 @@
+	}
+	const once_layer = once(new_layer);
+
+	$: $pixi && once_layer();
```
Don't love this

pngwn commented on js/imageeditor/shared/layers/Layers.svelte:
```diff
@@ -0,0 +1,193 @@
+		getContext<EditorContext>(EDITOR_KEY);
+
+	const LayerManager = layer_manager();
+	let layers: LayerScene[] = [];
```
We may need to expose more than just the current layers because it causes issues with undo of the current layer isn't the layer that the original command was performed on. Either than or figure out some way for commands to keep track of the current layer at the time.

pngwn commented on js/imageeditor/shared/tools/Brush.svelte:
```diff
@@ -0,0 +1,219 @@
+	// $: selected_size =
+	// 	_sizes[_sizes?.findIndex((c) => c === default_size)] || _sizes[0];
+
+	let selected_size = default_size;
```
This doesn't work. Fix.

pngwn commented on js/imageeditor/shared/tools/Brush.svelte:
```diff
@@ -0,0 +1,219 @@
+
+{#if current_option === "color" && colors}
+	<div>
+		<BrushColor
```
We should be able to close this without pressing escape

pngwn commented on js/imageeditor/shared/tools/Brush.svelte:
```diff
@@ -0,0 +1,219 @@
+		/>
+	</div>
+{:else if current_option === "size" && sizes}
+	<BrushSize
```
Ditto

pngwn commented on js/imageeditor/shared/tools/BrushSize.svelte:
This needs a size swatch

pngwn commented on js/imageeditor/shared/tools/Crop.svelte:
Need to add crop constraints. We'll need to be make this draggable for this to work.

pngwn commented on js/imageeditor/shared/tools/Sources.svelte:
```diff
@@ -0,0 +1,194 @@
+
+<style>
+	.modal {
+		position: fixed;
```
This probably isn't going to work on spaces. We'll need to do something more dynamic there using position absolute and the current scroll position. 

Probably a good time to fix the toasts positions because it is the same problem.

pngwn commented on js/imageeditor/shared/tools/Tools.svelte:
```diff
@@ -0,0 +1,160 @@
+	<Toolbar show_border={false}>
+		{#if sub_menu}
+			{#each sub_menu.options as meta (meta.id)}
+				<IconButton
+					highlight={sub_tool === meta.id}
+					on:click={meta.cb}
+					Icon={meta.icon}
+					size="large"
+					padded={false}
+				/>
```
We need to add the concept of a text overlay here so we can display the current brush size over the icon. The text will probably need to be in a store so we can make sure it always stays up to date. 

This is arguably the worst part of the code atm. Svelte 5 is gonna make this a million times nicer.

pngwn commented on js/imageeditor/shared/tools/brush.ts:
```diff
@@ -0,0 +1,246 @@
+ * @param point2 the second point
+ * @returns an array of points between the two points
+ */
+function interpolate(
+	point1: { x: number; y: number },
+	point2: { x: number; y: number }
+): Points[] {
+	let points: Points[] = [];
+	const dx = point2.x - point1.x;
+	const dy = point2.y - point1.y;
+	const distance = Math.sqrt(dx * dx + dy * dy);
+	const steps = Math.ceil(distance / 2);
+	const stepX = dx / steps;
+	const stepY = dy / steps;
+
+	for (let j = 0; j < steps; j++) {
+		const x = point1.x + j * stepX;
+		const y = point1.y + j * stepY;
+		points.push({ x, y });
+	}
+
+	return points;
+}
```
This seems to cause performance issues with a large brush size. Not 100% sure why but a simply limiting per frame in the handler might be enough.

pngwn commented on js/imageeditor/shared/tools/brush.ts:
```diff
@@ -0,0 +1,246 @@
+				paths.push({ x, y });
+			}
+
+			renderer.render(graphics, {
+				renderTexture:
+					mode === "draw" ? layer.draw_texture : layer.erase_texture,
+				clear: false
+			});
```
Not sure if it is the interpolation that is causing issues or just rendering large textures.

pngwn commented on js/imageeditor/shared/tools/sources.ts:
```diff
@@ -0,0 +1,115 @@
+	return {
+		start() {
+			container.removeChildren();
+			const gfx = make_graphics(0);
```
Naming consistency

pngwn commented on js/imageeditor/shared/utils/commands.ts:
```diff
@@ -0,0 +1,141 @@
+ * Creates a command manager
+ * @returns a command manager
+ */
+export function command_manager(): CommandManager {
```
We need a way to clear the history when we reset.

pngwn commented on js/imageeditor/shared/utils/pixi.ts:
```diff
@@ -0,0 +1,294 @@
+	 * The pixi container for masking
+	 */
+	mask_container: Container;
+	destroy(): void;
```
JSdoc for this.

pngwn commented on js/imageeditor/shared/utils/shaders.ts:
Remove.

pngwn commented on js/upload/src/Upload.svelte:
```diff
@@ -63,7 +64,15 @@
 		const target = e.target as HTMLInputElement;
 		if (!target.files) return;
-		await load_files(Array.from(target.files));
+		if (format != "blob") {
```
Update readme.

abidlabs commented:
Amazing work @pngwn! I just gave the image editor a quick spin and noticed a few potential bugs / UI suggestions:

1. When you're starting out, seeing all 7 buttons grouped together can be a bit overwhelming. I was thinking we could show only the 1st row of upload/webcam/paste buttons initially, and then show the others when you've provided an image. But I don't think that would work since you can sketch on a blank canvas without having provided any image. What if we moved the 2nd row of buttons to the side, in a vertical column underneath the "X" button. I feel like that might be a bit more intuitive than the current setup. (cc @hannahblair for your thoughts!)

2. Clicking the webcam icon opens a full-screen webcam capture. I assume not intentional?

3. The crop works beautifully. But clicking the rotate image icon has no effect.

4. Similarly, I can't get sketching or erasing working for me. Clicking the palette icon or the gray button has no effect and does not allow me to sketch on the base image

<img width="223" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/95b44f7b-f29e-4354-a683-f33b43338540">

5. Running `demo/image_editor/run.py` and clicking the "Run" button does not do anything for me. I see this error in the console: `Uncaught Error: Could not create blob`

pngwn commented:
1. I'd rather we didn't split the buttons up, you have to cover a lot of screen to get from one to the other, especially on larger screen sizes. Likewise on mobile having core controls in two different places will be irritating. It might be less visually overwhelming but it would be a worse UX. Buttons at the top are ones that are uses less frequently.

We could just start off with the brush selected, that way we have fewer buttons. We could even start off with nothing selected but again i think that is worse UX even though it might be OK visually. I don't really think users will be overwhelmed by 6 buttons though, most image editors have dozens on display at all times.

2. It is intentional but given your response, I'll tweak that.

3. I am going to remove rotate for now and implement later. Its complex (not the rotate, the UX).

4. I'll fix sketching / erasing. There is a bug atm (well 2). SO you have to crate a new layer and change the size of the brush for it to work. Should be sorted shortly.

5. Probably need to handle empty layers more gracefully. Will check that.

pngwn commented:
2. should be fixed now.

abidlabs commented on js/app/component_loader.js:
```diff
@@ -18,6 +19,7 @@ export async function load_component({ api_url, name, id, variant }) {
 			component: c
 		};
 	} catch (e) {
+		console.error(e);
```
Remove console error? Or maybe not

pngwn replied:
We need to keep this one in so users can understand what might have one wrong for debugging purposes.

abidlabs commented:
Good stuff @pngwn! I was able to get it running locally just fine. Here are the issues that I noticed:

1. When you open the color palette, there doesn't seem to be a way to close it. Clicking outside the palette didn't close it:

<img width="296" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/6ddcd1ae-7efe-4e25-8eb4-37d9b74bac34">

2. Similarly, when choosing a custom color, I would have expected that clicking on the color (the purple here) would have closed the palette, but it did not:

<img width="251" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/e2821b9f-6bad-4111-981f-1b9c0b0e7ac7">

3. The "X" icon to clear the image has no effect

4. I don't think the image data is being preprocessed correctly. Neither of these two demos work (no image output is shown when you click submit):

```py
import gradio as gr

demo = gr.Interface(lambda x:x, gr.ImageEditor(), gr.ImageEditor())

if __name__ == "__main__":
    demo.launch()
```

```py
import gradio as gr

demo = gr.Interface(lambda x:x["composite"], gr.ImageEditor(), gr.Image())

if __name__ == "__main__":
    demo.launch()
```

Some UI improvements would be nice (though we can do these in a separate PR afterwards):

* the brush radius selector

<img width="371" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/d1c73e45-be99-4c0c-a40c-0ea995607f77">

* not a huge fan of the two rows of buttons -- I think it creates too much unused white space at the bottom. I don't have a good alternative right now, perhaps @hannahblair or @gary149 do?

<img width="468" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/3846d7e3-181c-479c-ade9-11ee789dcc7d">


Cropping, brushing, layers, erasing, undo, and redo all work super well!

pngwn commented:
@abidlabs I've fixed all of those issues now I think. Also fixed the `change` and `input` events.

The issue with interface wasn't the preprocessing but the static variant.

aliabid94 commented:
Incredible work. Playing around with this, the few points of friction I hit:
1. Like abidlabs said, closing the popups for the brush / color controls was a little annoying
2. Hadn't immediately realized that uploading an image resets the entire history. I was trying to upload a second image on another layer, and then realized that uploading sets that as the new source entirely. (not something we need to change, just what I encountered)
3. Once I've selected a color and brush size, it would be nice if my cursor became a circle with that size and color. Would make it a little easier to know what I'm about to draw.
4. Brush sizes show weird numbers:
![Screen Shot 2023-11-17 at 11 35 58 AM](https://github.com/gradio-app/gradio/assets/7870876/25980468-fe0e-4a7e-b849-772145d66e03)
5. Python doesn't work on 3.8 (think this may be what you added @abidlabs)

pngwn commented:
> Brush sizes show weird numbers
That's pretty hilarious. Will remove the default size swatches for now. Not sure how useful they are.

pngwn commented:
I've tightened up the buttons a little now too. I have some ideas about keeping them one on line but will discuss first + try in another PR.

pngwn commented:
I think I broke the button background. Will check that. 

I'll take a look at those other issues too. Dark mode looks a bit janky. Will fix.

abidlabs commented:
Like the tightened buttons a lot. Noticed some small things:

1. It'd be good to know see visually what the boundaries of the canvas are for sketchpad inputs before you start sketching (perhaps a faint gray background on the areas that are outside of the canvas):

<img width="647" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/1d2a0076-cc93-4461-920d-8b3efcb25dd6">

2. When you first see the ImageEditor, its not clear what you're supposed to do:

<img width="659" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/4fbc710b-21f0-4970-b6e2-22192cb1c5d8">

Can we put a "drop image here" text similar to `gr.Image()` when you're on the uploading step

<img width="787" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/8c0b09b6-4885-4f97-912d-f9e36e557e4a">

abidlabs commented:
Removed some console logs so that we could run the static checks without linter complaining

pngwn commented on js/imageeditor/shared/ClearImage copy.svelte:
```diff
@@ -0,0 +1,30 @@
+<script lang="ts">
```
This file shouldn't even exist. Will delete. ty!

pngwn commented:
@abidlabs 

1. Added background when there is no background image
2. Added some text when there is no bg + no history + we are in bg mode. We cant add the drop text because dropping doesn't work atm.
3. I think that could be annoying if you want to upload a new image or something. After you leave the upload option you can't re-upload without clearing. If a `crop_size` is set by the author it will automatically go into crop though.
4. Will look into this
5. Is this when the ImageEditor is an output?

abidlabs commented:
For (5), yes when it is an output

pngwn commented:
@abidlabs @aliabid94 @hannahblair 

Thanks for the reviews!

I _think_ i've addressed everything.

@hannahblair I added some edge collisions to fix the issues on mobile (although mobile needs looking at on its own) and I tweaked darkmode so the color picker looks much nicer. I've also fixed the icon button (i did break them).

@aliabid94 I've cleaned up the UX regarding controls and remove those brush size related issues. I've also added a 'pain' cursor that matches the brush size as a preview for both erasing and drawing. Everything should be working on 3.8 now too (thanks @abidlabs ).

@abidlabs I've added a message for the blank state to give some pointers. We can probably improve that in the future. I've also fixed the statustracker and added the color preview on the circle.

The main new thing is you can now pass 'crop constraints' via `crop_size` and this will force the crop to match that ratio on the frontend and be resized fully on the backend.

abidlabs commented:
Doing some testing right now. Generally everything is working great. Just very tiny nits that I've found so far:

1. I would have expected clicking the "Upload an image or select..." box to open the upload file dialog like it does for our other components.

2. When you use an ImageEditor as an input and trigger the event (e.g. click a submit button), it takes quite a while before anything actually happens.

E.g. running

```py
import gradio as gr

gr.Interface(lambda x:x, "imageeditor", "imageeditor").launch()
```

I have to wait ~5 seconds before even just seeing the progress status tracker on the output component. 

3. Pressing the "X" button does not bring back the "Upload an image or select..." message. It just shows an empty white box for me (not sure if this is intentional):

<img width="389" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/df49ed49-7ca9-44ee-bb45-d67310359752">

4. "Clearing" an ImageEditor does not seem to work. Running the same demo as above:

```py
import gradio as gr

gr.Interface(lambda x:x, "imageeditor", "imageeditor").launch()
```

if you upload an image and then press the gray "Clear" button in the Interface, nothing happens.

abidlabs commented on gradio/components/image_editor.py:
```diff
@@ -0,0 +1,292 @@
+        mirror_webcam: bool = True,
+        show_share_button: bool | None = None,
+        _selectable: bool = False,
+        data_mode: Literal["image", "pathline"] = "image",
```
I don't think this is supported at the moment, right? Should we remove this param

abidlabs commented on gradio/components/image_editor.py:
```diff
@@ -0,0 +1,292 @@
+        _selectable: bool = False,
+        data_mode: Literal["image", "pathline"] = "image",
+        crop_size: tuple[int | float, int | float] | str | None = None,
+        transforms: Iterable[Literal["crop", "rotate"]] = ("crop", "rotate"),
```
"rotate" is not supported. Maybe we remove this as well for now?

abidlabs reviewed (approved):
I was only able to find small issues, so giving this the green check. Amazing work @pngwn!

pngwn commented:
I'll take a look at 1,2,3,3 and 5. 

6. I don't think we should remove transforms entirely but I'll change the default. 

7. Should already be possible by setting `color_mode` in `Brush` to `"fixed"`. 

8. I thought I fixed this, will double check. It might need a resize to take effect.

pngwn commented:
Okay, made a few changes.

I couldn't add click to upload yet due to how things are put together, we can do that later if it is really needed but feels off for an image editor imo.

- I've fixed up the 'empty' text so it is more appropriate depending what tools are shown
- Empty text now shows correctly when you clear the component
- Clearing works properly whether its done vis the x button or the `Clear` button
- Crop can now correctly be disabled.
- Eraser is only enabled if brush is enabled
- After uploading an image, crop will automatically be selected if a crop_size is set
- If `sources=[]` the draw tool will be selected.
- The status tracker now show on the uploading component (the editor) when we are preparing the data (uploading the files). Then after that is complete the normal process kicks off. We can improve the UX for this later but it should address the current lack of feedback (although 5 seconds seems a long time, i couldn't repro that).
- I've also cleaned up the python API a little. Previously it wasn't possible to disable the brush or eraser (because `None` cases get a default brush), so `eraser` and `brush` can now accept a boolean too.

pngwn commented:
And we are done!

Thanks @abidlabs @hannahblair @freddyaboulton @aliabid94  for the help with various bits and pieces, the PRs, and the detailed reviews + tested. Very much appreciated!

wamiq-reyaz commented:
Hey all, 

thanks for the wonderful work. I am still wondering about a few things though
1. The brushes are not well documented. Like how does one create a custom swatch?
2. If one instead uses a colorpicker component, how does one update the brush color for an ImageEditor? The current events do NOT target the brush color.
3. The documentation  for the ImageEditor does not actually showcase how the ImageEditor is actually used but instead has an unrelated rotation example
